kube-core/src/labels.rs

```diff
@@ -0,0 +1,552 @@
+                write!(
+                    f,
+                    "{key} in ({})",
+                    values.iter().cloned().collect::<Vec<_>>().join(",")
+                )
+            }
+            Expression::NotIn(key, values) => {
+                write!(
+                    f,
+                    "{key} notin ({})",
+                    values.iter().cloned().collect::<Vec<_>>().join(",")
+                )
+            }
+            Expression::Equal(key, value) => write!(f, "{key}={value}"),
+            Expression::NotEqual(key, value) => write!(f, "{key}!={value}"),
+            Expression::Exists(key) => write!(f, "{key}"),
+            Expression::DoesNotExist(key) => write!(f, "!{key}"),
+            Expression::Invalid => Ok(()),
+        }
+    }
+}
+
+impl Display for Selector {
+    /// Convert a selector to a string for the API
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let selectors: Vec<String> = self.0.iter().map(|e| e.to_string()).collect();
+        write!(f, "{}", selectors.join(","))
+    }
+}
+// convenience conversions for Selector and Expression
+
+impl IntoIterator for Expression {
+    type IntoIter = IntoIter<Self::Item>;
+    type Item = Self;
+
+    fn into_iter(self) -> Self::IntoIter {
+        Some(self).into_iter()
+    }
+}
+
+impl IntoIterator for Selector {
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+    type Item = Expression;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl FromIterator<(String, String)> for Selector {
+    fn from_iter<T: IntoIterator<Item = (String, String)>>(iter: T) -> Self {
+        Self::from_map(iter.into_iter().collect())
+    }
+}
+
+impl FromIterator<(&'static str, &'static str)> for Selector {
+    fn from_iter<T: IntoIterator<Item = (&'static str, &'static str)>>(iter: T) -> Self {
+        Self::from_map(
+            iter.into_iter()
+                .map(|(k, v)| (k.to_string(), v.to_string()))
+                .collect(),
+        )
+    }
+}
+
+impl FromIterator<Expression> for Selector {
+    fn from_iter<T: IntoIterator<Item = Expression>>(iter: T) -> Self {
+        Self::from_expressions(iter.into_iter().collect())
+    }
+}
+
+impl From<Expression> for Selector {
+    fn from(value: Expression) -> Self {
+        Self(vec![value])
+    }
+}
+
+impl From<LabelSelector> for Selector {
+    fn from(value: LabelSelector) -> Self {
+        let expressions = match value.match_expressions {
+            Some(requirements) => requirements.into_iter().map(Into::into).collect(),
+            None => vec![],
+        };
+        let mut equality: Selector = value
+            .match_labels
+            .map(|labels| labels.into_iter().collect())
+            .unwrap_or_default();
+        equality.0.extend(expressions);
+        equality
+    }
+}
+
+impl From<LabelSelectorRequirement> for Expression {
+    fn from(requirement: LabelSelectorRequirement) -> Self {
```

Comment:
Is there a particular reason you are avoiding more standard rust error handling here for invalid expressions?

Does it make more sense to make this a `TryFrom` with a new `SelectorError` representing the potential `Expression::Invalid` case?

Reply:
It makes questionable the `LabelSelector` implementation of `Matcher`, because instead of what is given should match, and the potential of error is in “what is given”, current implementation makes it so that error (which I believe can’t be logged within `kube-core`) converges to `false` match. I’m starting to think that implementing `Matcher` for `LabelSelector` by default is no longer a good idea.

Reply:
Yeah, I think you are right there. I put a follow-up comment on that implementation below.